File: daily-temps/cpp/daily_temperatures.cpp
#include "daily_temperatures.h"

#include <stack>
#include <vector>

std::vector<int> dailyTemperatures( const std::vector<int> &temperatures )
{
    std::vector<int> result( temperatures.size(), 0 );
    std::stack<int>  s;
    for( int i = 0; i < temperatures.size(); ++i ) {
        while( !s.empty() && temperatures[i] > temperatures[s.top()] ) {
            int prev = s.top();
            s.pop();
            result[prev] = i - prev;
        }
        s.push( i );
    }
    return result;
}

std::vector<int> dailyTemperaturesOpt( const std::vector<int> &temperatures )
{
    std::vector<int> res( temperatures.size(), 0 );
    std::vector<int> track;
    track.reserve( temperatures.size() );

    for( int i = 0; i < temperatures.size(); ++i ) {
        int currTemp = temperatures[i]; // <- preventing CPU-cache misses here
        while( !track.empty() && currTemp > temperatures[track.back()] ) {
            int prev = track.back();
            track.pop_back();
            res[prev] = i - prev;
        }
        track.push_back( i );
    }
    return res;
}

File: daily-temps/go/daily_temperatures.go
package dailytemps

func DailyTemperatures(temperatures []int) []int {
	result := make([]int, len(temperatures))
	var stack []int

	for i, temp := range temperatures {
		for len(stack) > 0 && temp > temperatures[stack[len(stack)-1]] {
			prevIndex := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			result[prevIndex] = i - prevIndex
		}
		stack = append(stack, i)
	}

	return result
}

func DailyTemperaturesOpt(temperatures []int) []int {
	n := len(temperatures)
	result := make([]int, n)
	stack := make([]int, 0, n/4)

	for i := 0; i < n; i++ {
		curr := temperatures[i]
		for len(stack) > 0 {
			top := stack[len(stack)-1] // safely assign once
			if curr <= temperatures[top] {
				break
			}
			stack = stack[:len(stack)-1]
			result[top] = i - top
		}
		stack = append(stack, i)
	}
	return result
}
